### PR TITLE
GCW-3417 Basic GoodCity API V2 Api client

### PR DESCRIPTION
**Pull request by BboyStatix**
## Ticket Link
https://jira.crossroads.org.hk/browse/GCW-3417
## What does this PR do?
Creates a basic api client implementation to communicate with GoodCity Api V2.
For now it implements the `/auth/send_pin` and `/auth/verify` endpoints.

### CODE REVIEW

File: src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      }
+    );
+  });
+
+  describe("Successful API response", () => {
+    it("it should return the appropriate response", async () => {
+      mockFetch.mockResolvedValueOnce({
+        ok: true,
+        json: async () => Promise.resolve(sendPinResponse),
+      } as Response);
```
Comment (BboyStatix) [inline src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts:33]: Is it really ok to satisfy the type-checking in this way? What about other important fields like `status`? Should we look into a fetch mocking third party library to abstract away the tedium?

File: src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import sendPinResponse from "./__mocks__/sendPinResponse.js";
+import verifyResponse from "./__mocks__/verifyResponse.js";
+
+const mockFetch = jest.spyOn(window, "fetch");
+
+afterEach(() => mockFetch.mockClear());
+
+afterAll(() => mockFetch.mockRestore());
+
+describe("sendPin", () => {
```
Comment (BboyStatix) [inline src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts:11]: Any ideas on a better structure for the tests? The current structure leads to a lot of duplication when new endpoints are added.
First of all we want to assert each method calls fetch correctly.
Secondly, we want to assert on the different type of responses which include
1. Successful API response. (we need to assert on the data format)
2. Error response from server e.g. (422 Mobile number invalid)
3. Network Error (e.g. request failed to reach server)

Among the above 1 and 2 are endpoint specific whereas Network Error is generic.
For 1 and 2 we need to create mocks and type definitions for the success case and error case. Any way to make this easier (e.g. autogeneration)?

File: src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import GoodCityApiV2Client from "./GoodCityApiV2Client";
```
Comment (patrixr) [inline src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.test.ts:1]: The general consensus on naming is that capitalised names refer to Types/Classes/Modules.

Here`GoodCityApiV2Client` is a function, so `goodCityApiV2Client` would be more appropriate.

I would say that the name is very long and verbose, `client` would be my preferred choice personally.

File: src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        type: data.type ?? "InternalServerError",
+        message: data.error ?? "Something went wrong",
+      });
+    }
+
+    return data;
+  });
+};
+
+export default GoodCityApiV2Client;
```
Comment (patrixr) [inline src/lib/client/GoodCityApiV2Client/GoodCityApiV2Client.ts]: #### Few things regarding this client:

- It does a hardcoded `POST`, that'll need to be configurable for other requests
- It enforces a body
- It doesn't support custom headers
- The base url is not configurable (not a big deal)
- There's a mix of promises and async code (you can use async everywhere and avoid `then`)
- The name is a bit off (wrote another comment about that somewhere). Even though it's named `goodcity` there's nothing in there that relates to goodcity apart from the baseUrl (which could be configurable). So why not name it "json client" ?

#### Potential options

- Make it into a class, with get/post/delete/patch/put methods
- Simply make multiple functions (get/post/etc)
- Make it into an entry point which returns the methods for all actions (more functional)

**Review by patrixr — CHANGES_REQUESTED** [review]

**Review by patrixr — APPROVED** [review]

**Review by bharat619 — APPROVED** [review]

**Pull request merged**